server/sonar-server/src/main/java/org/sonar/server/computation/ws/QualityGateStatusAction.java

```diff
@@ -0,0 +1,167 @@
+import org.sonar.api.server.ws.WebService;
+import org.sonar.batch.protocol.GsonHelper;
+import org.sonar.core.permission.GlobalPermissions;
+import org.sonar.core.util.Uuids;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.ce.CeActivityDto;
+import org.sonar.db.component.SnapshotDto;
+import org.sonar.db.measure.MeasureDto;
+import org.sonar.server.exceptions.NotFoundException;
+import org.sonar.server.user.UserSession;
+import org.sonarqube.ws.WsCe.QualityGateStatusWsResponse;
+import org.sonarqube.ws.WsCe.QualityGateStatusWsResponse.QualityGateDetails;
+import org.sonarqube.ws.client.ce.QualityGateStatusWsRequest;
+
+import static java.lang.String.format;
+import static org.sonar.server.ws.WsUtils.checkFound;
+import static org.sonar.server.ws.WsUtils.checkFoundWithOptional;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+
+public class QualityGateStatusAction implements CeWsAction {
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public QualityGateStatusAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    WebService.NewAction action = controller.createAction("quality_gate_status")
+      .setDescription("Quality gate status for a given CE task. <br />" +
+        "Returns a http code 404 if the snapshot associated with the task is not found.<br />" +
+        "Requires 'Administer System' or 'Execute Analysis' permission.")
+      .setResponseExample(getClass().getResource("quality_gate_status-example.json"))
+      .setSince("5.3")
+      .setInternal(true)
+      .setHandler(this);
+
+    action.createParam("taskId")
+      .setDescription("Task id")
+      .setRequired(true)
+      .setExampleValue(Uuids.UUID_EXAMPLE_03);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    QualityGateStatusWsRequest qualityGateStatusWsRequest = toQualityGateStatusWsRequest(request);
+    QualityGateStatusWsResponse qualityGateStatusWsResponse = doHandle(qualityGateStatusWsRequest);
+
+    writeProtobuf(qualityGateStatusWsResponse, request, response);
+  }
+
+  private QualityGateStatusWsResponse doHandle(QualityGateStatusWsRequest request) {
+    checkScanOrAdminPermission();
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      String taskUuid = request.getTaskId();
+      CeActivityDto ceActivityDto = getCeActivity(dbSession, taskUuid);
+      SnapshotDto snapshotDto = getSnapshot(dbSession, ceActivityDto.getSnapshotId(), taskUuid);
+      MeasureDto measureDto = getQualityGateDetailsMeasure(dbSession, snapshotDto, taskUuid);
+
+      return buildResponse(measureDto.getData(), taskUuid);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private CeActivityDto getCeActivity(DbSession dbSession, String taskUuid) {
```

Comment:
Important information: checking the quality gate is not mandatory on an analysis. It can be disabled. I don't know what's the status in this case (green or not set ?). Anyway I don't think HTTP code should be 404 in this case.